src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -130,6 +130,35 @@ private[concurrent] object Promise {
     override final def transformWith[S](f: Try[T] => Future[S])(implicit executor: ExecutionContext): Future[S] =
       dispatchOrAddCallbacks(get(), new Transformation[T, S](Xform_transformWith, f, executor))
 
+    override final def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+      val state = get()
+      if (state.isInstanceOf[Failure[T]]) {
+        this.asInstanceOf[Future[R]]
+      } else if (state.isInstanceOf[Success[T]]) {
+        val l = state.asInstanceOf[Success[T]].get
+        that.map(r => f(l, r))
+      } else {
+        val buffer = new AtomicReference[Success[Any]]()
+        val zipped = new DefaultPromise[R]()
+        this.onComplete {
+          case left: Success[T] =>
+            val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
+            if (right ne null)
+              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
+          case f: Failure[T] =>
+            zipped.tryComplete(f.asInstanceOf[Failure[R]])
+        }
+        that.onComplete {
+          case right: Success[U] =>
+            val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
+            if (left ne null)
+              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
+          case f: Failure[U] =>
+            zipped.tryComplete(f.asInstanceOf[Failure[R]])
+        }
+        zipped.future
+      }
+    }
```

Comment:
The following should be slightly faster:

```scala
override final def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val state = get()
      if (state.isInstanceOf[Try[T]]) {
        if (state.asInstanceOf[Try[T]].isFailure) this.asInstanceOf[Future[R]]
        else {
          val l = state.asInstanceOf[Success[T]].get
          that.map(r => f(l, r))
        }
      } else {
        val buffer = new AtomicReference[Success[Any]]()
        val zipped = new DefaultPromise[R]()

        val thisF: Try[T] => Unit = {
          case left: Success[T] =>
            val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
            if (right ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f => // Can only be Failure
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }

        val thatF: Try[U] => Unit = {
          case right: Success[U] =>
            val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
            if (left ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f => // Can only be Failure
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }
        // Cheaper than this.onComplete since we already polled the state
        this.dispatchOrAddCallbacks(state, new Transformation[T, Unit](Xform_onComplete, thisF, executor))
        that.onComplete(thatF)
        zipped.future
      }
    }
```

Reply:
@viktorklang I notice we have a fast path if the LHS `this` future in the `zip` has already completed; should we do fast paths for other combinations of things being already completed? e.g.

- LHS completed, RHS not
- RHS completed, LHS not
- RHS completed, LHS completed

We'd then only need to fall back to the `AtomicReference`/`DefaultPromise` implementation in the case where both futures aren't completed at time of `.zip`

Reply:
Something like this

```scala
    override final def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      (get(), that.value) match{
        case (_: Failure[T], _) => this.asInstanceOf[Future[R]]
        case (_, Some(_: Failure[U])) => that.asInstanceOf[Future[R]]
        case (t: Success[T], Some(r: Success[U])) => Future(f(t.value, r.value))
        case (null, None) =>
          val buffer = new AtomicReference[Success[Any]]()
          val zipped = new DefaultPromise[R]()

          val thisF: Try[T] => Unit = {
            case left: Success[T] =>
              val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
              if (right ne null)
                zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
            case f => // Can only be Failure
              zipped.tryComplete(f.asInstanceOf[Failure[R]])
          }

          val thatF: Try[U] => Unit = {
            case right: Success[U] =>
              val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
              if (left ne null)
                zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
            case f => // Can only be Failure
              zipped.tryComplete(f.asInstanceOf[Failure[R]])
          }
          // Cheaper than this.onComplete since we already polled the state
          this.onComplete(thisF)
          that.onComplete(thatF)
          zipped.future
      }
    }
```

Reply:
I'm not sure how much we care about the "already completed" fast paths here, but I figure if we're gonna fast path one special case we should fast path all three of them (especially given how little code is required to do so)